In [8]:
import pandas as pd

# Read the data into a Pandas DataFrame
url = "https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv"
data = pd.read_csv(url)
data.head()

# Make a prediction about which model will perform better (to be done in a markdown cell)

# Create the labels set (y) from the "spam" column
y = data["spam"]

# Create the features (X) DataFrame from the remaining columns
X = data.drop(columns=["spam"])

from sklearn.model_selection import train_test_split

# Split the data into training and testing datasets using train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1)

# Check the balance of the labels variable (y) using the value_counts function
y.value_counts()


spam
0    2788
1    1813
Name: count, dtype: int64

### Prediction:
I predict that the Random Forest model will outperform the Logistic Regression model in terms of accuracy. Random Forests combine multiple decision trees, which allows them to capture complex, non-linear relationships in the data. On the other hand, Logistic Regression assumes a linear relationship between the features and the target, which may limit its performance if the dataset includes non-linear patterns. However, the performance of each model will depend on the characteristics of the dataset.


In [9]:
from sklearn.preprocessing import StandardScaler

# Create an instance of StandardScaler
scaler = StandardScaler()

# Fit the Standard Scaler with the training data
X_train_scaled = scaler.fit_transform(X_train)

# Scale the testing features DataFrame using the transform function
X_test_scaled = scaler.transform(X_test)


In [10]:
from sklearn.linear_model import LogisticRegression

# Create a logistic regression model with random_state set to 1
lr_model = LogisticRegression(random_state=1)

# Fit the logistic regression model using the scaled training data
lr_model.fit(X_train_scaled, y_train)

# Save the predictions on the testing data labels using the testing feature data
lr_predictions = lr_model.predict(X_test_scaled)

from sklearn.metrics import accuracy_score

# Evaluate the model's performance by calculating the accuracy score of the model
lr_accuracy = accuracy_score(y_test, lr_predictions)
print(f"Logistic Regression Accuracy: {lr_accuracy}")


Logistic Regression Accuracy: 0.9261404779145547


In [11]:
from sklearn.ensemble import RandomForestClassifier

# Create a random forest classifier model with random_state set to 1
rf_model = RandomForestClassifier(random_state=1)

# Fit the random forest classifier model using the scaled training data
rf_model.fit(X_train_scaled, y_train)

# Save the predictions on the testing data labels using the testing feature data
rf_predictions = rf_model.predict(X_test_scaled)

# Evaluate the model's performance by calculating the accuracy score of the model
rf_accuracy = accuracy_score(y_test, rf_predictions)
print(f"Random Forest Accuracy: {rf_accuracy}")


Random Forest Accuracy: 0.9594496741491673


### Evaluation
Based on the accuracy scores, the results matched the prediction. The Random Forest model (0.9594) outperformed the Logistic Regression model (0.9261). This suggests that the Random Forest model is better suited for this dataset. Logistic Regression, being a linear model, works well when there is a linear relationship between features and the target variable. It is simpler and less prone to overfitting with clean data. However, it assumes linearity, which might not capture the complexities of the data. On the other hand, Random Forest, an ensemble learning method, constructs multiple decision trees and can capture non-linear relationships and interactions between features. It is more robust and less likely to overfit due to the averaging of multiple trees. The higher accuracy of Random Forest indicates the presence of non-linear relationships and complex feature interactions in the dataset, which Logistic Regression might miss. Overall, the Random Forest model's ability to handle non-linear relationships and robustness to overfitting contributed to its better performance, aligning with the prediction that it would perform better in this scenario.